hercules-ci-agent/hercules-ci-agent/Hercules/Agent/Config.hs

```diff
@@ -1,26 +1,38 @@
-{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 module Hercules.Agent.Config where
 
-import           Protolude
+import           Protolude               hiding ( to )
 import qualified System.Environment
-import qualified Dhall
+import           Toml
+
+data ConfigPath = TomlPath FilePath
 
 data Config = Config
-  { herculesApiBaseURL :: Text
+  { herculesApiBaseURL :: Maybe Text
   , clusterJoinTokenPath :: Text
   , concurrentTasks :: Integer
   , cacheKeysPath :: Maybe Text
-  } deriving (Generic, Dhall.Inject, Dhall.Interpret)
-
-newDefaultConfig :: IO Config
-newDefaultConfig = do
-  baseUrl <- determineDefaultApiBaseUrl
-  pure Config
-    { herculesApiBaseURL = baseUrl
-    , clusterJoinTokenPath = panic "Config.clusterJoinTokenPath wasn't set."
-    , concurrentTasks = 4
-    , cacheKeysPath = Nothing
-    }
+  } deriving (Generic)
+
+tomlCodec :: TomlCodec Config
+tomlCodec =
+  Config
+    <$> dioptional (Toml.text "apiBaseUrl")
+    .= herculesApiBaseURL
+    <*> Toml.text "clusterJoinTokenPath"
+    .= clusterJoinTokenPath
+    <*> Toml.integer "concurrentTasks"
+    .= concurrentTasks
+    <*> dioptional (Toml.text "cacheKeysPath")
+    .= cacheKeysPath
+
+defaultConfig :: Config
+defaultConfig = Config
```

Comment:
Could we set all options to be a `Maybe` and set defaults also for the TOML? Right now defaultConfig is not used, but Config should have defaults.

Reply:
join token path and concurrent tasks are actually required now, because the toml parser requires it. So technically only the rest is optional (or overridable)

Reply:
`concurrentTasks` should have a default for TOML, maybe that can be wired up in `readConfig`

Reply:
The default is hardcoded to `4` like in the module now.